File: android/app/src/main/java/com/henrydashwood/magpie/data/PodcastProgress.kt
package com.henrydashwood.magpie.data

import java.util.UUID

/** Immutable once sent: retry the exact ID/body after a lost acknowledgement. */
data class PodcastProgressReport(val expectedRevision: String, val seconds: Double, val completed: Boolean,
    val requestId: String = UUID.randomUUID().toString()) {
    init {
        require(expectedRevision.matches(Regex("[a-f0-9]{64}")))
        require(seconds.isFinite() && seconds >= 0)
        require(requestId.matches(Regex("[a-zA-Z0-9-]{1,64}")))
    }
}

/** A retry can acknowledge an older write while returning newer device/filing state. */
data class PodcastProgressReceipt(val episode: RemoteEpisode, val acceptedRevision: String) {
    val changedSinceAcceptance: Boolean get() = episode.progressRevision != acceptedRevision
}

interface PodcastProgressApi {
    suspend fun podcastProgress(token: String, episodeId: Int, report: PodcastProgressReport): PodcastProgressReceipt
}

File: ios/Hearful/API/PodcastProgress.swift
import Foundation

nonisolated struct PodcastProgressReport: Codable, Equatable, Sendable {
    let requestID: String
    let expectedRevision: String
    let seconds: Double
    let completed: Bool
    var isValid: Bool {
        isProgressToken(expectedRevision) && seconds.isFinite && seconds >= 0
            && (1...64).contains(requestID.utf8.count)
            && requestID.utf8.allSatisfy { (48...57).contains($0) || (65...90).contains($0) || (97...122).contains($0) || $0 == 45 }
    }
    enum CodingKeys: String, CodingKey {
        case completed
        case requestID = "request_id"
        case expectedRevision = "expected_revision"
        case seconds = "position_seconds"
    }
}

nonisolated struct PodcastProgressReceipt: Decodable, Sendable {
    let episode: Episode
    let acceptedRevision: String
    var changedSinceAcceptance: Bool { episode.progressRevision != acceptedRevision }
    enum CodingKeys: String, CodingKey {
        case episode
        case acceptedRevision = "accepted_revision"
    }
}
